numba/np/random/distributions.py

```diff
@@ -529,3 +529,165 @@ def random_poisson(bitgen, lam):
 def random_negative_binomial(bitgen, n, p):
     Y = random_gamma(bitgen, n, (1 - p) / p)
     return random_poisson(bitgen, Y)
+
+
+@register_jitable
+def random_binomial_btpe(bitgen, n, p):
```

Comment:
This particular function (which is implemented using `goto` statements in C in NumPy source code) leads to massive slowdown during compilation, and potentially needs to be optimized somehow.

Reply:
Hi @kc611 , the curiosity of seeing a bunch of goto was too interesting to skip. I took your approach, and then converted your code into a finite state machine using various approaches, but all of them take many minutes to compile. Your "while" and "if" based switch table is the winner for now.

Reply:
I have managed to decrease the compile time, but only by about 10%, by re-arranging your code. See [my fork](https://github.com/bszollosinagy/numba/commits/refactor_binomial_goto)

I have traced the algo to its original paper, where they discuss how efficient it will be at executing, but did not consider compilation times:
Kachitvichyanukul, V., & Schmeiser, B. W. (1988). Binomial random variate generation. Communications of the ACM, 31(2), 216-222.

Now sure how to improve on the compilation time any more than this. AFAIK the finite state machine is about the only acceptable use of GOTO nowadays, because aligns well with the problem domain (literally jumping around states).

Reply:
Hi, @bszollosinagy Thank you for for your interest in this PR. :smiley: 

> I took your approach, and then converted your code into a finite state machine using various approaches, but all of them take many minutes to compile. Your "while" and "if" based switch table is the winner for now.

Yep, same was the case for me. Of the all the logical approaches I took, this particular one seems to compile and execute the fastest, hence I drafted the PR with this one. 

The changes you made on your fork look pretty great. I think those might pretty much be the optimizations we could make with the 'while' + 'if' approach. Do you want to continue development on the issue ? If so, I'll leave this PR be for a while, so your fork can take over. I think the tests have already been written and the only work remaining would be making sure the types that Numba lowers the logic into are correct to prevent overflows.

Reply:
No probs, I'll adopt changes from your fork,, they look good. 

The type verification was required in previous such PRs to avoid using the wrong floating precision such as accidentally using 32 bit float in a 64 bit algorithmic implementation, since without intervention, Numba will assign the default precision bits, same for signed - unsigned floats.